[CP] LRSD-7729 Caching jira search results

jennyhchen opened the pull request:
https://liferay.atlassian.net/browse/LRSD-7729

- only re-caches the first page (I was debating if we should paginate and cache all the non-filtered result pages but by default, we'd only list the newer tickets anyway and not many people would really need to check every LSV; also any subsequent calls for the same search results (including filters) would grab the cached results on any user, let me know if you still want me to add it though)
- spring boot has a weird cache issue with their annotation, due to how their proxies work, (1) you can't cache a method within the same class as it only caches once at the start of a request and (2) it was caching the jwt which means every request would be unique due to the user making the call; so in order to cache only partner vs customer results, I moved it out so that it can also be used in the scheduler method to evict the caches by manually using CacheManager
- let me know if you do just want to update the caches once a day, I can update the cron (if any of you have a preferred time as well)
- also, do you think it might be beneficial to make the cache evict method accessible so we can manually run it (depending on when security team wants to release an LSV update)

liferay-continuous-integration commented:
To conserve resources, the PR Tester does not automatically run for every pull.

If your code changes were already tested in another pull, reference that pull in this pull so the test results can be analyzed.

If your pull was never tested, comment &quot;ci:test&quot; to run the PR Tester for this pull.

amosfong commented:
we're not really taking advantage of using spring's caching here. It could be replaced with a hashmap and it would be the same. This is what I suggest:

- Move the _search method to another class that takes in pre-built JQL and add @cacheable to it and let spring cache handle it
- instead of using now(), always round down to midnight of current day
- clear cache a little before midnight every day
- a clear cache endpoint might be useful to have
- I dont think we need to repopulate cache after clearing, just let recache naturally

There could be a day delay whenever they make changes on LSV's and if they are aware they can set their release dates always to be before midnight so it shows up the next day.

jennyhchen commented:
> * Move the _search method to another class that takes in pre-built JQL and add @Cacheable to it and let spring cache handle it

Yeah okay, I agree. The alternative option was moving some of the calls to another class, but we seemed to want to keep it all in one class for now so I went with that approach. I'll work on moving it and resend.

> * instead of using now(), always round down to midnight of current day
> * clear cache a little before midnight every day

What did you mean by rounding down?

amosfong commented:
> What did you mean by rounding down?

instead of `sb.append(" <= now()");`, do `sb.append(" <= " + [YYYY-MM-DD 00:00:00]);`

amosfong commented:
> If you do not add any specific cache library, Spring Boot auto-configures a [simple provider](https://docs.spring.io/spring-boot/reference/io/caching.html#io.caching.provider.simple) that uses concurrent maps in memory. When a cache is required (such as piDecimals in the preceding example), this provider creates it for you. The simple provider is not really recommended for production usage,...

We probably don't want to use the default cache provider. I dont think our case is complex so Caffeine seems like it's the simplest one out of the list:
https://www.baeldung.com/spring-boot-caffeine-cache

jennyhchen commented:
Just for future reference... when or what criteria should I look out for to know when to use a non-default cache provider? I remember you did mention trying to use a simple HashMap so I had originally thought Caffeine might be a good option but to keep it even more simple would be to just try using the default cache provider.

jennyhchen commented:
@amosfong 

>instead of sb.append(" <= now()");, do sb.append(" <= " + [YYYY-MM-DD 00:00:00]);

What would be the benefit of this? 🤔  I'm looking to see if this is possible and some people reported timezone issues if we try to round down to midnight. What time zone region would I need to set the check to? I think using the now() would have similar issues. Also, Release/Support/Security teams aren't very consistent with their updates. 

I can try to do some tests with CachePut and see if that might be a viable option but from the documentation, requests would still be sent to Jira which sounded like the load lag would still be present if we pushed this to prod.

amosfong commented:
> What would be the benefit of this? 

It would make cache consistent across pages. For example, if I hit page 1 now, and don't check page 2 until hours later, the results could be different

> What time zone region would I need to set the check to?

It doesn't really matter, we just need to be consistent. Probably whatever timezone JIRA is using would be best so the security team can know what to expect when setting their release dates.

> I can try to do some tests with CachePut and see if that might be a viable option but from the documentation, requests would still be sent to Jira which sounded like the load lag would still be present if we pushed this to prod.

What would we use cacheput for?

jennyhchen commented:
>It would make cache consistent across pages. For example, if I hit page 1 now, and don't check page 2 until hours later, the results could be different. It would also make caching more predictable in case we need to debug issues

Wouldn't that be the same as if Jira is in Japan Time and I'm working in Pacific Time and come back a few hours later and the date changed? There would be some inconsistencies either way but I also think it would be the same behavior even when I'm searching for tickets on Jira. I shouldn't expect the same page results if a new ticket has been published while I was gone.

Also, since we're signed in to the one Jira user that's making the requests, it might just be based on that user's timezone. There will still be hours of delay either way if that's the case.

>What would we use cacheput for?

To update the cache if there are any changes. Problem is the frontend loading lag so I want to test to see if that might be an issue. To get around that, I was then thinking of a spring boot event listener to update the caches.

amosfong commented:
> Wouldn't that be the same as if Jira is in Japan Time and I'm working in Pacific Time and come back a few hours later and the date changed? 

Not exactly. For example, User hits first page which has tickets LPS-1->20 and it gets cached. Later user hits page 2 and some new tickets have been published since then. LPS-1->20 is now in page 2 so it gets cached there. Now there are LPS's is on both page 1 and 2.

> Also, since we're signed in to the one Jira user that's making the requests, it might just be based on that user's timezone. There will still be hours of delay either way if that's the case.

If a delay is a problem, we can coordinate our cache clear times with security team's publish date times. We can double check with security team how strictly do publish times need to be respected.

> To update the cache if there are any changes. Problem is the frontend loading lag so I want to test to see if that might be an issue. To get around that, I was then thinking of a spring boot event listener to update the caches.

We aren't prefetching here. Let the cache naturally populate. The only exception I would say is the home page if we really wanted it

jennyhchen commented:
>Not exactly. For example, User hits first page which has tickets LPS-1->20 and it gets cached. Later user hits page 2 and some new tickets have been published since then. LPS-1->20 is now in page 2 so it gets cached there. Now there are LPS's is on both page 1 and 2.
>We aren't prefetching here. Let the cache naturally populate. The only exception I would say is the home page if we really wanted it

I was thinking of post-fetching, not pre-fetching. 

1. User sees LPS-1-20 on page 1
2. Cacheable - because it's the first search, method will request Jira and cache the results (unless we re-cache home page after doing a cache evict)
3. Event Listener when step 2 method is called or value is grabbed from cache (seeing if this is possible) -> update the cache
4. there may be a slight delay but at least, if the customer goes to page 2 and sees LPS-5 on page 2, if they go back to page 1, hopefully new page 1 results will be cached

^ Just an idea but I'll look into formatting the time and see if that helps with the consistency.

amosfong commented:
So if a user hits the cache, our CX will asynchronously update that cache entry? That seems unorthodox to me. Is this a common pattern you found somewhere? I'd say skip step #3

jennyhchen commented:
I was just thinking that we do similar stuff for our indexes in provisioning so was thinking if it's possible to do similar approaches for caches to keep them up to date rather than evicting them daily. (I don't think we need to keep it that complicated so I'll finish testing the `yyyy-MM-dd 00:00` approach to see if we get more consistent results that way.)

amosfong commented:
oh i see. These are two different things with different purposes. This is a cache while provisioning is a search index. Let's stick with a simple cache implementation and if we want to add stuff, we can do it later.

jennyhchen commented:
okay, re-pushed

- moved into a JiraService class in order to use caching annotations
- added "issue" cache for individual tickets and tested as well
- also moved the cron job to a listener class (added a rest controller method to call it manually too if we need to)
- added Caffeine max size of 1000 (number of entries)

amosfong commented on workspaces/liferay-customer-workspace/client-extensions/liferay-customer-etc-spring-boot/src/main/java/com/liferay/customer/JiraRestController.java:
```diff
@@ -42,14 +35,28 @@
+	@RequestMapping(method = RequestMethod.GET, path = "/jira/cacheEvict")
+	public ResponseEntity<String> cacheEvict() throws Exception {
+		try {
+			_cacheMessageListener.cacheEvictJira();
```
can we put a cacheEvict annotation here instead?

jennyhchen replied:
Locally, I tested before the permission changes just now and it does work. Did you want me to move it back?

I searched for best practices and it seems the reason we might want to keep it separate is to make sure if some listeners/jobs take a while to execute, it might clog up the traffic to the rest controller.

amosfong replied:
yea i think it'd be cleaner

jennyhchen replied:
I don't think I can anymore unless I didn't need to add the permissions, but we'd be exposing the endpoint (the cache evict annotation will run automatically). Thoughts? 🤔

jennyhchen replied:
I can also remove the `beforeInvocation = true` part if we're not going to recache the first pages. It was added because re-caching the first page didn't work unless the caches were evicted before the method ran. Thoughts on if we want to add it back or should I remove that part of the annotation?

amosfong replied:
yea remove the beforeInvocation. Let's not recache pages.

According to the docs, the method has to return successfully before it clears cache. So a permission exception would not clear any cache.

jennyhchen replied:
@Scheduled can't be used though if we do a permission check
` Initialization of bean failed; nested exception is java.lang.IllegalStateException: Encountered invalid @Scheduled method 'delete': Only no-arg methods may be annotated with @Scheduled`

amosfong replied:
the scheduled cache eviction can be a different method

jennyhchen replied:
So it can stay where it is in JiraService? We can't use cacheable annotations after the request to the controller (unless we proxy to another class).

https://github.com/liferay-customer-solutions/liferay-portal/pull/460/commits/c1d109ab0afc50814c0406b816674e309090eb56#diff-8f25ac9fb32fb822d5819c39797219a84dcca4fd49e58d6a6932da0da6c575a3R42-R45

amosfong replied:
yea it's good in jiraservice. I would actually prefer it there cause that centralizes the cache logic in one place

amosfong commented on workspaces/liferay-customer-workspace/client-extensions/liferay-customer-etc-spring-boot/src/main/java/com/liferay/customer/JiraRestController.java:
```diff
@@ -42,14 +35,28 @@
 @RestController
 public class JiraRestController extends BaseRestController {
 
+	@RequestMapping(method = RequestMethod.GET, path = "/jira/cacheEvict")
```
this isn't really RESTful. Maybe a DELETE on /jira/cache?

jennyhchen replied:
Sorry, I was doing some logging and forgot to update it. I updated and re-pushed. Also added an Admin check because I don't think we'd want other people to find that and mess with our caches.

amosfong commented on workspaces/liferay-customer-workspace/client-extensions/liferay-customer-etc-spring-boot/src/main/java/com/liferay/customer/service/JiraService.java:
```diff
@@ -0,0 +1,571 @@
+		sb.append(" = 'Ready for Publishing'");
+
+		DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern(
+			"yyyy-MM-dd");
```
can we add the constant time here?
"yyyy-MM-DD '00:00'"

jennyhchen replied:
it would still need the same SB lines, I can update it to
`DateTimeFormatter dateTimeFormatter = DateTimeFormatter.ofPattern(
			"yyyy-MM-dd 00:00");` though if you prefer that look

jennyhchen replied:
i updated it, let me know if you prefer something else `("yyyy-MM-DD '00:00'")` will break the json search, the whole date needs to be in single quotes or jira will throw errors

`project = 'LSV' AND cf[10929] = 'Ready for Publishing' AND cf[10927] <= '2025-01-15 00:00' ORDER BY cf[10927] desc, cf[10786] ASC`

amosfong replied:
ok, that's fine. It's just clearer to me if the whole dateformat comes from one place

amosfong commented on workspaces/liferay-customer-workspace/client-extensions/liferay-customer-etc-spring-boot/src/main/java/com/liferay/customer/JiraRestController.java:
```diff
@@ -132,141 +138,22 @@ public ResponseEntity<String> search(
-				_calculateStartAt(page, pageSize));
-
-			JSONObject responseJSONObject = _transformSearchResults(jsonObject);
+			JSONObject jsonObject = _jiraService.search(
```
can these be combined?
_jiraService.search(
					filterAffectedVersions, filterCategories,
					filterClassifications, filterFixVersions, filterSeverities,
					keywords, page, pageSize, sortOrder, hasEarlyPublishAccess(jwt));

jennyhchen replied:
yep, thanks for catching that, I was testing and didn't catch it. updated.

amosfong commented on workspaces/liferay-customer-workspace/client-extensions/liferay-customer-etc-spring-boot/src/main/java/com/liferay/customer/messaging/CacheMessageListener.java:
```diff
@@ -0,0 +1,25 @@
+		allEntries = true, beforeInvocation = true, value = {"issue", "issues"}
+	)
+	@Scheduled(cron = "0 0 0 * * *")
+	public void cacheEvictJira() throws Exception {
```
can this be moved to JiraService? I don't think it needs to be a separate class (it's not a messageListener either)

jennyhchen replied:
moved

jennyhchen replied:
I was originally thinking we might need to separate out our cron jobs/schedulers/message listeners similar to the ones we do on the other repos. Yeah, it's not really a message listener. I wasn't sure what to name it actually. I don't think some of our jobs were message listeners either on provisioning/customer since they just run every 15 minutes or once a day similar to a cron but they are extending the BaseMessageListener class.

When would we need to separate out jobs and schedulers and what would we name it? (e.g. if we have LFU, we'd need to run a possible job to delete old files)

amosfong commented:
made a few changes:
https://github.com/liferay-customer-solutions/liferay-portal/pull/472

-switched affectedversions to new caching method
-avoid verbs in REST url paths

nice work on figuring this out
